### ISSUE TITLE
PreferredLifetime = 0 isn't setting the lifetime to 0

### ISSUE THREAD

**Issue by anitazha** [systemd/systemd#20891]
**systemd version the issue has been seen with**
 > 249.4

I've been using the config below:

```
$ cat /etc/systemd/network/1-eth0.network 

[Network]
Address = <primary address>
Address = <secondary address>
IPv6AcceptRA = true

[Match]
MACAddress = <mac address>
Name = eth0

[IPv6AcceptRA]
UseAutonomousPrefix = false
UseOnLinkPrefix = false
DHCPv6Client = false

[Address]
Address = <secondary address>
PreferredLifetime = 0
```

If I do `ip addr` and the secondary address has preferred lifetime set to forever, running `networkctl reload && networkctl reconfigure eth0` on systemd 248.5 would change the lifetime to 0. However in systemd 249.4 this isn't changing to 0.

**Comment by DaanDeMeyer** [comment]
If we look at the documentation for the Address property in the Network section, we find the following:

> A static IPv4 or IPv6 address and its prefix length, separated by a "/" character. Specify this key more than once to configure several addresses. The format of the address must be as described in inet_pton(3). This is a short-hand for an [Address] section only containing an Address key (see below). This option may be specified more than once.

So what you're doing here is configuring the same address twice, once with the default preferred lifetime and once with the preferred lifetime set to 0. Internally, networkd stores addresses in a ordered hashmap by the config file section (and filename). So in practice, the same address appears twice in that hashmap and gets configured twice. In this case, the address with preferred lifetime 0 appears before the same address with preferred lifetime forever in the hashmap, meaning it gets configured with lifetime zero and immediately after gets configured with lifetime forever again.

So I think this is working as intended (not sure if we still need to keep the old behavior working as well for backwards compat). It should suffice to simply remove the `Address = <secondary address>` line from the [Network] config section to fix the issue.

**Comment by yuwata** [comment]
Hm? the two `<secondary address>` in the above is the same address?

BTW, at least, current HEAD does not have the reported issue. See #20892. But, note that yesterday I merged slightly big PR which changes how addresses are internally managed by networkd.

**Comment by DaanDeMeyer** [comment]
Yes, they're the same address.

Should we warn on the same address appearing in an [Address] section and the [Network] section?

**Comment by yuwata** [comment]
One more note, even if the `<secondary address>` are equivalent, the issue should NOT happen.
As described in the above, each addresses are stored in ordered hashmap, so, first the address is configured with lifetime _forever_, then later re-configured _lifetime=0_.

**Comment by DaanDeMeyer** [comment]
I have some logging (not sure if I can share, will need to ask Anita) that shows the opposite happening, first the secondary address is configured with lifetime 0, afterwards its reconfigured with lifetime forever.

**Comment by yuwata** [comment]
> Should we warn on the same address appearing in an [Address] section and the [Network] section?

Yeah, maybe. We should dedup sections before configuring addresses, routes, and so on. But that's may not be simple. As, we need to implement the exactly same way what kernel uses. And, sometimes, kernel changes that...

**Comment by yuwata** [comment]
> I have some logging (not sure if I can share, will need to ask Anita) that shows the opposite happening, first the secondary address is configured with lifetime 0, afterwards its reconfigured with lifetime forever.

Interesting.

**Comment by DaanDeMeyer** [comment]
```
eth0: Requesting address: <secondary-address>(valid forever, preferred for 0), flags: n/a
eth0: Requesting address: <secondary-address> (valid forever, preferred forever), flags: n/a
```

**Comment by yuwata** [comment]
Indeed... Strange.

As you can see the log in `NetworkdNetworkTests.test_address_static` in `systemd-networkd-tests.py` running on CentOS CIs, all Address= settings or [Address] sections are processed in order.

**Comment by DaanDeMeyer** [comment]
@anitazha Can you verify the config sections were in the same order as you posted them here when you were testing this?

**Comment by yuwata** [comment]
Anyway, workaround is simple. Please drop the redundant `Address=` in [Network] section. (I am happy that the issue is not so critical than I initially expected, hehe)

**Comment by DaanDeMeyer** [comment]
I'll make a PR to warn when users do this

**Comment by anitazha** [comment]
> Scrubbed relevant log lines:
> 
> ```
> eth0: Requesting address: <secondary-address>(valid forever, preferred for 0), flags: n/a
> eth0: Requesting address: <secondary-address> (valid forever, preferred forever), flags: n/a
> eth0: Requesting address: <primary-address> (valid forever, preferred forever), flags: n/a
> ```
> 
Ah so these logs I sent Daan they were gotten with `journalctl -r` so they're reversed! The addresses are processed in order. 

But the odd thing is later on in the logs I see:
`eth0: Forgetting address: <secondary-address> (valid forever, preferred forever), flags: permanent`
with no further mentions of `preferred for 0` anymore.

Anyways I tried dropping the first `Address=<secondary address>` from `[Network]` and it works so I can work around that. I can also try with the HEAD and see if the issue exists (since I saw https://github.com/systemd/systemd/pull/20892 doesn't have this issue)

**Comment by anitazha** [comment]
Okay it was really bothering me because I think this should work; the later address specification should definitely override the first when they have the same address, right? Since the config is processed top to bottom.

main branch seems to have the issue still. I saw the word "dedup" in 40b12fa20dcf8ae3138de4384a869a298a69ee19 and thought it looked suspicious. So I tested the config with it and with the commit before it. It's definitely 40b12fa20dcf8ae3138de4384a869a298a69ee19. @yuwata is this the correct handling when the same address is supplied multiple times?
